packages/phone-number-privacy/signer/src/pnp/services/quota.ts

```diff
@@ -0,0 +1,156 @@
+import { ContractKit } from '@celo/contractkit'
+import {
+  ErrorMessage,
+  PnpQuotaRequest,
+  PnpQuotaStatus,
+  SignMessageRequest,
+} from '@celo/phone-number-privacy-common'
+import { Knex } from 'knex'
+import { ACCOUNTS_TABLE } from '../../common/database/models/account'
+import { REQUESTS_TABLE } from '../../common/database/models/request'
+import { getPerformedQueryCount, incrementQueryCount } from '../../common/database/wrappers/account'
+import { storeRequest } from '../../common/database/wrappers/request'
+import { Counters, Histograms, meter } from '../../common/metrics'
+import { OdisQuotaStatusResult, QuotaService } from '../../common/quota'
+import { getBlockNumber } from '../../common/web3/contracts'
+import { config } from '../../config'
+import { PnpSession } from '../session'
+
+export abstract class PnpQuotaService
+  implements QuotaService<SignMessageRequest | PnpQuotaRequest> {
+  protected abstract readonly requestsTable: REQUESTS_TABLE
+  protected abstract readonly accountsTable: ACCOUNTS_TABLE
+
+  constructor(readonly db: Knex, readonly kit: ContractKit) {}
+
+  public async checkAndUpdateQuotaStatus(
+    state: PnpQuotaStatus,
+    session: PnpSession<SignMessageRequest>,
+    trx: Knex.Transaction
+  ): Promise<OdisQuotaStatusResult<SignMessageRequest>> {
+    const remainingQuota = state.totalQuota - state.performedQueryCount
+    Histograms.userRemainingQuotaAtRequest.labels(session.request.url).observe(remainingQuota)
+    let sufficient = remainingQuota > 0
+    if (!sufficient) {
+      session.logger.warn({ ...state }, 'No remaining quota')
+      if (this.bypassQuotaForE2ETesting(session.request.body)) {
+        Counters.testQuotaBypassedRequests.inc()
+        session.logger.info(session.request.body, 'Request will bypass quota check for e2e testing')
+        sufficient = true
+      }
+    } else {
+      await Promise.all([
+        storeRequest(
+          this.db,
+          this.requestsTable,
+          session.request.body.account,
+          session.request.body.blindedQueryPhoneNumber,
+          session.logger,
+          trx
+        ),
+        incrementQueryCount(
```

Comment:
I didn't have enough time to be 100% sure, but I don't see any mitigation to the attack where a verified account can fund quota for many wallet addresses. Is this being handled?

Reply:
Is the legacy signer quota system intended to be shut down as a final mitigation?

Reply:
Yes that's our plan. Once Komenci is no longer sending us traffic we'll be free to shut that down